File: engines/web_comics.engine/src/sound-manager.js
/**
 * Sound Manager
 * Handles sound playback based on scroll position
 */

export class SoundManager {
  constructor(archive, comics, enabled = true) {
    this.archive = archive;
    this.comics = comics;
    this.enabled = enabled;

    this.audioContext = null;
    this.audioBuffers = new Map();
    this.activeSources = new Map();
    this.playedOneShots = new Set();
    this.lastScrollY = 0;

    this._initAudioContext();
  }

  _initAudioContext() {
    // Lazy init on first user interaction
    if (typeof window !== 'undefined') {
      const initOnInteraction = () => {
        if (!this.audioContext) {
          this.audioContext = new (window.AudioContext || window.webkitAudioContext)();
        }
        document.removeEventListener('click', initOnInteraction);
        document.removeEventListener('touchstart', initOnInteraction);
      };

      document.addEventListener('click', initOnInteraction);
      document.addEventListener('touchstart', initOnInteraction);
    }
  }

  async _loadSound(src) {
    if (this.audioBuffers.has(src)) {
      return this.audioBuffers.get(src);
    }

    const soundPath = `sounds/${src}`;
    const data = this.archive[soundPath];

    if (!data) {
      console.warn(`Sound not found: ${soundPath}`);
      return null;
    }

    if (!this.audioContext) {
      this.audioContext = new (window.AudioContext || window.webkitAudioContext)();
    }

    try {
      const audioBuffer = await this.audioContext.decodeAudioData(data.buffer.slice(0));
      this.audioBuffers.set(src, audioBuffer);
      return audioBuffer;
    } catch (e) {
      console.error(`Failed to decode audio: ${src}`, e);
      return null;
    }
  }

  async _playSound(sound) {
    if (!this.enabled || !this.audioContext) return;

    const buffer = await this._loadSound(sound.src);
    if (!buffer) return;

    // Resume audio context if suspended
    if (this.audioContext.state === 'suspended') {
      await this.audioContext.resume();
    }

    const source = this.audioContext.createBufferSource();
    source.buffer = buffer;

    // Apply volume
    const gainNode = this.audioContext.createGain();
    gainNode.gain.value = sound.volume;

    source.connect(gainNode);
    gainNode.connect(this.audioContext.destination);

    source.loop = sound.loop || false;
    source.start(0);

    // Track active sources for looping sounds
    if (sound.loop) {
      this.activeSources.set(sound.src, { source, gainNode });
    }

    source.onended = () => {
      if (!sound.loop) {
        // Allow replay on scroll back
        // this.playedOneShots.delete(sound.src);
      }
    };
  }

  _stopSound(src) {
    const active = this.activeSources.get(src);
    if (active) {
      try {
        active.source.stop();
      } catch (e) {
        // Already stopped
      }
      this.activeSources.delete(src);
    }
  }

  /**
   * Process sounds based on scroll position
   * @param {number} scrollY - Current scroll position
   */
  process(scrollY) {
    if (!this.enabled || !this.comics.sounds) return;

    const scrollingDown = scrollY > this.lastScrollY;
    this.lastScrollY = scrollY;

    for (const sound of this.comics.sounds) {
      if (sound.type === 'point') {
        // One-shot sound at specific scroll position
        if (scrollingDown &&
            scrollY >= sound.start &&
            !this.playedOneShots.has(sound.src)) {
          this.playedOneShots.add(sound.src);
          this._playSound(sound);
        }
      } else if (sound.type === 'range') {
        // Range sound - play while in range
        const inRange = scrollY >= sound.start &&
                       (sound.end === null || scrollY <= sound.end);
        const isPlaying = this.activeSources.has(sound.src);

        if (inRange && !isPlaying) {
          this._playSound({ ...sound, loop: true });
        } else if (!inRange && isPlaying) {
          this._stopSound(sound.src);
        }
      }
    }
  }

  setEnabled(enabled) {
    this.enabled = enabled;

    if (!enabled) {
      // Stop all active sounds
      this.activeSources.forEach((_, src) => this._stopSound(src));
    }
  }

  pause() {
    this.audioContext?.suspend();
  }

  resume() {
    this.audioContext?.resume();
  }

  destroy() {
    this.activeSources.forEach((_, src) => this._stopSound(src));
    this.audioBuffers.clear();
    this.playedOneShots.clear();

    if (this.audioContext) {
      this.audioContext.close();
      this.audioContext = null;
    }
  }
}

export default SoundManager;

File: engines/unity_comics.engine/Runtime/Audio/SoundManager.cs
using System;
using System.Collections.Generic;
using UnityEngine;
using NativeMind.ComicsViewer.Models;
using NativeMind.ComicsViewer.Core;

namespace NativeMind.ComicsViewer.Audio
{
    /// <summary>
    /// Manages scroll-triggered sound playback
    /// </summary>
    public class SoundManager : IDisposable
    {
        private readonly IComicsSource _source;
        private readonly Comics _comics;
        private readonly GameObject _audioRoot;
        private readonly Dictionary<string, AudioSource> _audioSources;
        private readonly Dictionary<string, AudioClip> _audioClips;
        private readonly HashSet<string> _playedOneShots;

        private bool _enabled;
        private float _lastScrollY;

        public SoundManager(IComicsSource source, Comics comics, bool enabled = true)
        {
            _source = source;
            _comics = comics;
            _enabled = enabled;

            _audioRoot = new GameObject("SoundManager");
            _audioSources = new Dictionary<string, AudioSource>();
            _audioClips = new Dictionary<string, AudioClip>();
            _playedOneShots = new HashSet<string>();

            PreloadSounds();
        }

        private void PreloadSounds()
        {
            if (_comics.sounds == null) return;

            foreach (var sound in _comics.sounds)
            {
                LoadSound(sound.src);
            }
        }

        private AudioClip LoadSound(string src)
        {
            if (_audioClips.TryGetValue(src, out var cachedClip))
            {
                return cachedClip;
            }

            string soundPath = _source.GetSoundPath(src);
            if (string.IsNullOrEmpty(soundPath) || !System.IO.File.Exists(soundPath))
            {
                Debug.LogWarning($"Sound not found: {src}");
                return null;
            }

            // Load audio file
            // Note: For production, use proper async loading with UnityWebRequest
            var audioData = System.IO.File.ReadAllBytes(soundPath);
            var clip = CreateAudioClipFromWav(audioData, src);

            if (clip != null)
            {
                _audioClips[src] = clip;
            }

            return clip;
        }

        private AudioClip CreateAudioClipFromWav(byte[] data, string name)
        {
            // Simple WAV parser - for production use a proper audio library
            try
            {
                // Skip WAV header (44 bytes for standard WAV)
                int headerSize = 44;
                int sampleCount = (data.Length - headerSize) / 2;

                // Assume 16-bit, mono, 44100 Hz
                var clip = AudioClip.Create(name, sampleCount, 1, 44100, false);

                float[] samples = new float[sampleCount];
                for (int i = 0; i < sampleCount; i++)
                {
                    int offset = headerSize + i * 2;
                    if (offset + 1 < data.Length)
                    {
                        short sample = (short)(data[offset] | (data[offset + 1] << 8));
                        samples[i] = sample / 32768f;
                    }
                }

                clip.SetData(samples, 0);
                return clip;
            }
            catch (Exception e)
            {
                Debug.LogError($"Failed to parse audio: {name} - {e.Message}");
                return null;
            }
        }

        private AudioSource GetOrCreateSource(string src)
        {
            if (_audioSources.TryGetValue(src, out var source))
            {
                return source;
            }

            var clip = LoadSound(src);
            if (clip == null) return null;

            source = _audioRoot.AddComponent<AudioSource>();
            source.clip = clip;
            source.playOnAwake = false;

            _audioSources[src] = source;
            return source;
        }

        /// <summary>
        /// Process sounds based on scroll position
        /// </summary>
        public void Process(float scrollY)
        {
            if (!_enabled || _comics.sounds == null) return;

            bool scrollingDown = scrollY > _lastScrollY;
            _lastScrollY = scrollY;

            foreach (var sound in _comics.sounds)
            {
                ProcessSound(sound, scrollY, scrollingDown);
            }
        }

        private void ProcessSound(Sound sound, float scrollY, bool scrollingDown)
        {
            var source = GetOrCreateSource(sound.src);
            if (source == null) return;

            source.volume = sound.volume;

            if (sound.type == "point")
            {
                // One-shot at specific scroll position
                if (scrollingDown &&
                    scrollY >= sound.start &&
                    !_playedOneShots.Contains(sound.src))
                {
                    _playedOneShots.Add(sound.src);
                    source.loop = false;
                    source.Play();
                }
            }
            else if (sound.type == "range")
            {
                // Play while in range
                bool inRange = scrollY >= sound.start &&
                              (sound.end < 0 || scrollY <= sound.end);

                if (inRange && !source.isPlaying)
                {
                    source.loop = sound.loop;
                    source.Play();
                }
                else if (!inRange && source.isPlaying)
                {
                    source.Stop();
                }
            }
        }

        public void SetEnabled(bool enabled)
        {
            _enabled = enabled;

            if (!enabled)
            {
                StopAll();
            }
        }

        public void Pause()
        {
            foreach (var source in _audioSources.Values)
            {
                if (source.isPlaying)
                {
                    source.Pause();
                }
            }
        }

        public void Resume()
        {
            foreach (var source in _audioSources.Values)
            {
                source.UnPause();
            }
        }

        public void StopAll()
        {
            foreach (var source in _audioSources.Values)
            {
                source.Stop();
            }
        }

        public void Dispose()
        {
            StopAll();

            foreach (var clip in _audioClips.Values)
            {
                if (clip != null)
                {
                    UnityEngine.Object.Destroy(clip);
                }
            }

            _audioClips.Clear();
            _audioSources.Clear();

            if (_audioRoot != null)
            {
                UnityEngine.Object.Destroy(_audioRoot);
            }
        }
    }
}
